### PR TITLE
[ticket/15692] Move checks if file exists from adapter to storage

### PR DESCRIPTION
**Pull request by rubencm**
Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

todo: do the renames with storage in the migration to update it in the table too

Depends on #4933 

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15692

### CODE REVIEW

**Comment by rubencm** [comment]
!set WIP :construction:

**Comment by Nicofuma** [comment]
looks good

**Comment by rubencm** [comment]
This PR go together with #5247 but i did it in two parts so check the other better

**Comment by rubencm** [comment]
!set WIP :construction:

**Comment by rubencm** [comment]
!unset WIP :construction:

**Comment by rubencm** [comment]
!unset WIP :construction:

File: phpBB/includes/functions_download.php [hunk 1]
```diff
@@ -223,9 +223,6 @@ function send_file_to_browser($attachment, $category)
 		}
 	}
 
-	// Close the db connection before sending the file etc.
-	file_gc(false);
```
Comment (Nicofuma) [inline phpBB/includes/functions_download.php]: why moving it?

File: phpBB/phpbb/storage/storage.php [hunk 2]
```diff
@@ -131,22 +142,28 @@ public function get_contents($path)
 	 *
 	 * @param string	$path	file/directory to check
 	 *
-	 * @return bool	Returns true if the file/directory exist, false otherwise.
+	 * @return bool	Returns true if the file/directory exist, false otherwise
 	 */
-	public function exists($path)
+	public function exists($path, $full_check = false)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: you should update the method comment with the new parameter

File: phpBB/phpbb/storage/storage.php [hunk 3]
```diff
@@ -256,7 +317,10 @@ public function track_file($path, $update = false)
 
 		if (!$row)
 		{
-			$file = $this->file_info($path);
+			// Don't call the method, because it check's if the file is tracked
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: I guess it shouls be `Don't call the file_info method,`

File: phpBB/phpbb/storage/storage.php [hunk 4]
```diff
@@ -295,6 +359,30 @@ public function untrack_file($path)
...
+	 * @return bool	True if file is tracked
+	 */
+	public function is_tracked($path)
+	{
+		$sql_ary = array(
+			'file_path'		=> $path,
+			'storage'		=> $this->get_name(),
+		);
+
+		// Get file, if exist update filesize, if not add new record
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: wrong comment

File: phpBB/includes/functions_download.php (hunk 1 above)
Comment (rubencm) [reply phpBB/includes/functions_download.php]: Read_stream use database to check if the file exist

**Pull request merged**